### PR TITLE
Handle keeping previous OSX versions more smoothly on Mac

### PR DESCRIPTION
**Pull request by thboop**
Outstanding questions
- Should we add back in deleting old runner versions in this pr or the next one?
- Need to validate these commands are all available in the base images of osx, they have man pages so they should be

### CODE REVIEW

File: src/Misc/layoutbin/update.sh.template [hunk 1]
```diff
@@ -118,40 +118,47 @@ then
...
-        cp "$rootfolder/externals/node12/bin/node" "$rootfolder/externals.2.280.2/node12/bin/node"
+        date "+[%F %T-%4N] Failed to find runner pgid" >> "$logfile" 2>&1
+        mv -fv "$logfile" "$logfile.failed"
+        exit 1
     fi
 
-    if [[ ! -e "$rootfolder/externals.2.280.1/node12/bin/node" ]]
+    # inspect the open file handles to find the node process
+    # we can't actually inspect the process using ps because it uses relative paths and doesn't follow symlinks
+    path=$(lsof -a -g $procgroup -F n | grep node12/bin/node | grep externals. | tail -1 | cut -c2-)
```
Comment (pje) [inline src/Misc/layoutbin/update.sh.template:140]:
```suggestion
    path=$(lsof -a -g $procgroup -F n | grep node12/bin/node | grep externals | tail -1 | cut -c2-)
```

File: src/Misc/layoutbin/update.sh.template [hunk 2]
```diff
@@ -118,40 +118,47 @@ then
...
-        mkdir -p "$rootfolder/externals.2.280.1/node12/bin"
-        cp "$rootfolder/externals/node12/bin/node" "$rootfolder/externals.2.280.1/node12/bin/node"
+        date "+[%F %T-%4N] Failed to find runner path" >> "$logfile" 2>&1
+        mv -fv "$logfile" "$logfile.failed"
+        exit 1
     fi
 
-    if [[ ! -e "$rootfolder/externals.2.279.0/node12/bin/node" ]]
+    # trim the last 5 characters of the path '/node'
+    trimmedpath=$(echo ${path::$((${#path} - 5))})
```
Comment (pje) [inline src/Misc/layoutbin/update.sh.template:149]: Can we just use `dirname` for this instead?


```suggestion
    trimmedpath=$(dirname "$path")
```

**Review by pje — DISMISSED** [review]

File: src/Misc/layoutbin/update.sh.template [hunk 3]
```diff
@@ -118,40 +118,47 @@ then
...
-    # we will recreate all the ./externals/node12/bin/node of the past 5 versions
-    # v2.280.3 v2.280.2 v2.280.1 v2.279.0 v2.278.0
-    if [[ ! -e "$rootfolder/externals.2.280.3/node12/bin/node" ]]
-    then
-        mkdir -p "$rootfolder/externals.2.280.3/node12/bin"
-        cp "$rootfolder/externals/node12/bin/node" "$rootfolder/externals.2.280.3/node12/bin/node"
-    fi
-
-    if [[ ! -e "$rootfolder/externals.2.280.2/node12/bin/node" ]]
+if [[ "$currentplatform" == 'darwin'  && restartinteractiverunner -eq 0 ]]; then
```
Comment (TingluoHuang) [inline src/Misc/layoutbin/update.sh.template:123]: should we rely on `restartinteractiverunner` or `-f $rootfolder/.service`?

Comment (thboop) [reply src/Misc/layoutbin/update.sh.template:123]: spoke in person, lets use restartinteractiverunner for now

File: src/Misc/layoutbin/update.sh.template [hunk 4]
```diff
@@ -118,40 +120,102 @@ then
     exit 1
 fi
 
-# fix upgrade issue with macOS
+# fix upgrade issue with macOS when running as a service
+attemptedtargetedfix=0
+currentversion=$(defaults read loginwindow SystemVersionStampAsString)
```
Comment (thboop) [inline src/Misc/layoutbin/update.sh.template:125]: this needs to be moved inside the if statement

**Review by pje — DISMISSED** [review]
👍 as a hotfix. (In the near future we should invest some time in cleaning up the runner entrypoint/upgrade code!)

File: src/Misc/layoutbin/update.sh.template [hunk 5]
```diff
@@ -18,6 +18,8 @@ downloadrunnerversion=_DOWNLOAD_RUNNER_VERSION_
 logfile="_UPDATE_LOG_"
 restartinteractiverunner=_RESTART_INTERACTIVE_RUNNER_
 
+telemetryfile=".telemetry"
```
Comment (TingluoHuang) [inline src/Misc/layoutbin/update.sh.template:21]: can we easily move this to `_diag` folder? if not, I am totally fine with here.

File: src/Misc/layoutbin/update.sh.template [hunk 6]
```diff
@@ -118,40 +120,104 @@ then
...
-    # we will recreate all the ./externals/node12/bin/node of the past 5 versions
-    # v2.280.3 v2.280.2 v2.280.1 v2.279.0 v2.278.0
-    if [[ ! -e "$rootfolder/externals.2.280.3/node12/bin/node" ]]
+if [[ "$currentplatform" == 'darwin'  && restartinteractiverunner -eq 0 ]]; then
+    # We needed a fix for https://github.com/actions/runner/issues/743
+    # We will recreate the ./externals/node12/bin/node of the past runner version that launched the runnerlistener service
+    # Otherwise mac gatekeeper kills the processes we spawn on creation as we are running a process with no backing file
+
+    # get version for telemetry
+    currentversion=$(defaults read loginwindow SystemVersionStampAsString)
```
Comment (TingluoHuang) [inline src/Misc/layoutbin/update.sh.template:132]: Do we need this? we know the current version from Kusto.

will `defaults read loginwindow SystemVersionStampAsString` ever error?

File: src/Misc/layoutbin/update.sh.template [hunk 7]
```diff
@@ -118,40 +120,104 @@ then
...
+                date "+[%F %T-%4N] DarwinRunnerUpgrade: Failed to trim runner path. TrimmedPath: $trimmedpath, path: $path, pgid: $procgroup, root: $rootfolder, version: $currentversion" >> "$logfile" 2>&1
+                date "+[%F %T-%4N] DarwinRunnerUpgrade: Failed to trim runner path. TrimmedPath: $trimmedpath, path: $path, pgid: $procgroup, root: $rootfolder, version: $currentversion" >> "$telemetryfile" 2>&1
+            fi
+        else
+            date "+[%F %T-%4N] DarwinRunnerUpgrade: Failed to find runner path. Path: $path, pgid: $procgroup, root: $rootfolder, version: $currentversion" >> "$logfile" 2>&1
+            date "+[%F %T-%4N] DarwinRunnerUpgrade: Failed to find runner path. Path: $path, pgid: $procgroup, root: $rootfolder, version: $currentversion" >> "$telemetryfile" 2>&1
+        fi
+    else
+            date "+[%F %T-%4N] DarwinRunnerUpgrade: Failed to find runner pgid. pgid: $procgroup, root: $rootfolder, version: $currentversion" >> "$logfile" 2>&1
+            date "+[%F %T-%4N] DarwinRunnerUpgrade: Failed to find runner pgid. pgid: $procgroup, root: $rootfolder, version: $currentversion" >> "$telemetryfile" 2>&1
```
Comment (TingluoHuang) [inline src/Misc/layoutbin/update.sh.template:168]: what about we just store the `$logfile` to `$telemetryfile`?

Comment (thboop) [reply src/Misc/layoutbin/update.sh.template:168]: I don't think we need all of that information, trying to minimize how much we capture.

File: src/Misc/layoutbin/update.sh.template (hunk 6 above)
Comment (thboop) [reply src/Misc/layoutbin/update.sh.template:132]: 👍 I'll remove it

File: src/Runner.Worker/JobRunner.cs [hunk 8]
```diff
@@ -285,6 +289,30 @@ private void MaskTelemetrySecrets(List<JobTelemetry> jobTelemetry)
...
+                if (File.Exists(telemetryFilePath))
+                {
+                    var telemetryData = File.ReadAllText(telemetryFilePath, Encoding.UTF8);
+                    var telemetry = new JobTelemetry
+                    {
+                        Message = $"Runner File Telemetry:\n{telemetryData}",
+                        Type = JobTelemetryType.General
+                    };
+                    jobTelemetry.Add(telemetry);
+                    File.Delete(telemetryFilePath);
```
Comment (TingluoHuang) [inline src/Runner.Worker/JobRunner.cs:306]: Try `IOUtil.DeleteFile()`

**Review by TingluoHuang — APPROVED** [review]

**Review by aminaka-ui — COMMENTED** [review]
> @cdb #554 `
Duplicate of #1104

**Pull request merged**